### ISSUE TITLE
Set Servo options triggered by failsafe

### ISSUE THREAD

**Issue by alexanderever** [ArduPilot/ardupilot#23927]
## Feature request

For UAS which perform a function controlled by Set_servo outputs, such as crop spraying, it would be desirable for the UAS to stop doing the function if a failsafe mode is triggered. It is not clear if the Sprayer functionality includes this but it may be useful for other  applications such as vectored thrust VTOLs etc

Currently the consequence of RC signal loss while spraying would be the UAS performing a RTL while still spraying. This would be VERY BAD as the spray would be released at a high altitude and subsequently drift to unintended locations. 

**Describe the solution you'd like**
_A clear and concise description of what you want to happen._

A parameter to set servo outputs should failsafe be triggered Eg SERVOXX_FS_VALUE = 1000 

**Describe alternatives you've considered**
_A clear and concise description of any alternative solutions or features you've considered._
Lua scripting?

checked the Sprayer functionality in ardupilot, it is unclear if this issue is addressed. 

(Also documentation on "SPRAY_SPINNER Spinner rotation speed: Spinner's rotation speed in PWM (a higher rate will disperse the spray over a wider area horizontally)" is inaccurate and could cause serious issues, increased RPM actually creates smaller droplets.  see https://en.wikipedia.org/wiki/Rotary_atomizers)

**Platform**
[ x ] All
[x  ] AntennaTracker
[ x ] Copter
[  x] Plane
[ x ] Rover
[  x] Submarine

**Additional context**
_Add any other context or screenshots about the feature request here._

**Comment by rmackay9** [comment]
Thanks for the feature request.  Turning off the sprayer during any kind of failsafe makes a lot of sense.

**Comment by alexanderever** [comment]
Hi Randy,

thanks for considering the feature!

 I work for an agricultural company with a long history of spraying in ground based and aerial settings. We also contribute to international discussions on drones as spraying systems. There are a number of features we can think of that we feel would improve Arducopter equipped aircraft for this role.

i'd be very interested in helping the development of Arducopter.

Please feel free to contact me at ***@***.***

Thanks,

Alexander

________________________________

**Comment by peterbarker** [comment]
@alexanderever a script has been offered up to accomplish what you wanted here.  Would you be happy with such a solution?

**Comment by alexanderever** [comment]
Hi Peter, 

Its great that my idea has been considered. thanks to everyone involved! 

A script is a good start, and i think anyone serious about making a spraying drone with arducopter should have the ability to implement a script if they are aware of it! perhaps it could be mentioned in the "crop sprayer" documentation. 

However i think that as you are able to enable sprayer from parameters, the risk that someone builds a UASS without really knowing what they're doing and then accidentally destroying a neighbour's crop is fairly high. For this reason i think it would be best if the logic to stop spraying on failsafe was in the main code. (is spray_enable available on rover?)

The scope of my initial feature request was broader than just the spray functionality though. i believe the ability to set an output value on GPIO that is reverted to when a failsafe is triggered would be a valuable feature for many types of vehicle. Examples could include lowering gear or retracting mounts, triggering the ballast system on Sub etc.

**Comment by rmackay9** [comment]
I agree with @alexanderever that we should just add it as a SPRAYER_OPTIONS parameter bit because this request has come up so many times.

**Comment by Priyanshu-Shah** [comment]
Hi @rmackay9 @alexanderever,

PR opened #31771 

I wanted to make my first contribution here, based on the conversation, I have made the following changes:

1. New SPRAY_OPTION param
2. When enabled, entering failsafe forces sprayer off and keeps it off while failsafe is active (and requires re-enabling after failsafe clears, this was a design choice, please let me know your views on this).
3. Currently called on for radio and gcs failsafes. 
4. Verified locally with mavproxy and sitl. 

Kindly let me know any changes/suggestions/additions to be made.

**Comment by alexanderever** [comment]
Hi Priyanshu, 

Thank you for taking the time to do this! 

From my perspective the logic you have implimented makes perfect sense (and i think that it should be enabled as default).

With your current changes does the sprayer stop if battery failsafe is triggered? If the UAS went into RTL for any reason, the sprayer should stop to prevent a) the release of droplets from high altitude causing drift or b) the UAS spraying as it returns, either spraying an unintended area or overdosing an already treated area. 

Another potential safety feature could be to limit the altitude at which the sprayer can be enabled.  

Many thanks

Alexander

**Comment by Priyanshu-Shah** [comment]
Sure thing, I was looking for a green flag before moving on with other failsafe conditions. The height factor makes sense too but how do you propose it to be added? Hard coded into different levels, left for user input, or potentially a sensor based approach (but I fear it would complicate things too much), let me know your thoughts on this.

**Comment by alexanderever** [comment]
I think it would be best to look for input from Randy or Peter on the altitude limit. but i think its likely that most drones used for spraying will have a downward facing rangefinder. If you were to impliment this limit id suggest having it as a max spraying altitude parameter that the user can enter in meters. it is possible that someone has a usecase for spraying at altitudes which would usually be undesirable

**Comment by Priyanshu-Shah** [comment]
I have added the battery and RTL failsafe as well, it just turns off the servos if battery levels reach critical or when any type of RTL action is performed for any failsafe. 

For clarity, not all failsafes are critical, so currently, sprayer turns off if:
- RC/GCS failsafe
- Battery Critical failsafe
- Any type of failsafe but the action is RTL based or landing based. 

@rmackay9 @peterbarker Requesting feedback on the height based cap. What kind of an implementation is suggested?

**Comment by shashwatvoa** [comment]
I just made a pull request adding this fearure can someone review it and guide me as I am new here

**Comment by RafiShaik-AI** [comment]
Hello, I am new to contributing to ArduPilot and interested in working on this issue.
Could you please confirm if this issue is suitable for a beginner contributor or suggest where to start?
Thank you.
